### ISSUE TITLE
allow ordering and grouping of subcommands in help output

### ISSUE THREAD

**Issue by jclulow** [trentm/node-cmdln#5]
For complicated tools with a rich set of subcommands, the help output is often best understood when related commands are grouped together.  It would be good to allow the caller to provide a subcommand sort order, including group headings, for use in printing the help message.  For example:

```javascript
#!/usr/bin/env node
/* vim: set ts=8 sts=8 sw=8 noet: */

var mod_util = require('util');
var mod_cmdln = require('/ws/node-cmdln');


function
TestCommand()
{
        mod_cmdln.Cmdln.call(this, {
                name: 'testcmd',
                desc: 'This is a Test Command',
                subcmdOrder: [
                        { group: 'Most Excellent Commands' },
                        'awesome',
                        { group: 'Other Commands', unmatched: true }
                ]
        });
}
mod_util.inherits(TestCommand, mod_cmdln.Cmdln);

TestCommand.prototype.do_awesome = function
do_awesome(subcmd, opts, args, cb)
{
        console.log('Do Awesome.');
        cb();
};
TestCommand.prototype.do_awesome.help = 'Do awesome things.';
TestCommand.prototype.do_awesome.aliases = [ 'awe', 'a' ];

TestCommand.prototype.do_other = function
do_other(subcmd, opts, args, cb)
{
        console.log('Do Other.');
        cb();
};
TestCommand.prototype.do_other.help = 'Do other things.';

mod_cmdln.main(new TestCommand());
```

The above program would present a help message that looks like this:

```
$ ./testcmd.js
This is a Test Command

Usage:
    testcmd [OPTIONS] COMMAND [ARGS...]
    testcmd help COMMAND

Options:
    -h, --help      Show this help message and exit.

Commands:
  Most Excellent Commands:
    awesome (awe, a)  Do awesome things.

  Other Commands:
    help (?)        Help on a specific sub-command.
    other           Do other things.

```

If the consumer does not provide the new `subcmdOrder` directive, the current behaviour would be preserved.  This array property, if provided, contains a mixture of:

* subcommand names; e.g. `"awesome-command"` for `Program#do_awesome_command()`
* group headings; e.g. `{ group: "A Few Of My Favourite Things" }`

The consumer may provide _at most one_ group heading object that also contains the property `unmatched: true`; this group will include all subcommands found through reflection that are not explicitly called out in the sort order.  To omit the `unmatched` group when providing a sort order that does not explicitly include every enumerated subcommand is a programming error and will throw.  To include a subcommand in the sort order for which a handler is not found via reflection is also a programming error.

**Comment by trentm** [comment]
Yes, thanks!  Two things:

1. May I propose that you changethe field name to `helpSubcmds` instead of `subcmdOrder`? For a couple reasons:

(a) That'll fit with the other `help*` options to the constructor. 

(b) That allows this to eventually (separate commit) be the mechanism for other customizations of the top-level help output. E.g. for `imgadm help` I override `.printHelp` entirely so the "Commands: ..." block can have a short synopsis for each subcommand instead of just the command name:

```
...
Commands:
    imgadm help [<command>]                help on commands

    imgadm sources [<options>]             list and edit image sources

    imgadm avail                           list available images
    imgadm show <uuid|docker-repo-tag>     show manifest of an available image

    imgadm import [-P <pool>] <image-id>   import image from a source
    imgadm install [-P <pool>] -m <manifest> -f <file>
...
```

I could see eventually specifying that synopsis there:

```
function
TestCommand()
{
        mod_cmdln.Cmdln.call(this, {
                name: 'imgadm',
                desc: 'manage thy images',
                subcmdOrder: [
                        { group: 'Most Excellent Commands' },
                        { name: 'install', synopsis: 'imgadm install [-P <pool>] -m <manifest> -f <file>' },
                        { group: 'Other Commands', unmatched: true }
                ]
        });
}
```

2. Does `{ group: "" }` work if I just want separation (as in the imgadm example)? I remember asking the same thing for the equivalent in dashdash.

**Comment by jclulow** [comment]
I have changed the configuration property to be `helpSubcmds`, and made sure that a blank group heading will result in a single blank line being printed.

**Comment by trentm** [comment]
Merged and in cmdln@3.1.0. Thanks!
